EWB-4381: Cleanup

Lax125 opened the pull request:
# Description

Clean up code:
- Missing `@JvmOverloads` added to non-abstract constructors with default parameters
- Make test classes and methods `internal`
- Go through Kotlin issues in code inspection
- Fix typos and grammar mistakes in documentation

# Associated tasks

Tasks blocking this one:
- [x] https://github.com/zepben/evolve-sdk-jvm/pull/158

Tasks this is blocking:
- None

# Checklist

If any of these are not applicable, strikethrough the line `~like this~`. **Do not delete it!**. Let the reviewer decide if you should have done it.

### Code
- [x] I have performed a self review of my own code (including checking issues raised when creating the PR).
- [x] I have added/updated unit tests for these changes, and if not I have explained why they are not necessary.
- [x] I have commented my code in any hard-to-understand or hacky areas.
- [x] I have handled all new warnings generated by the compiler or IDE.
- [x] I have rebased onto the target branch (usually main).
      
### Documentation
- [x] I have updated the changelog.
- [x] I have updated any documentation required for these changes.

# Breaking Changes
- [x] I have considered if this is a breaking change and will communicate it with other team members if so.

* Removed `IdentifiedObject.removeNamesFromTypes()`. Use `IdentifiedObject.clearNames()` instead.
* Removed `DiagramServiceInstanceCache` and `NetworkServiceInstanceCache`.

ryanjcoleman commented:
Task linked: [EWB-4381 Clean up SDK](https://app.clickup.com/t/6929263/EWB-4381)

Lax125 commented:
Base changed so you can easily see what the fixes are in the diff.

Lax125 commented on src/main/kotlin/com/zepben/evolve/database/sqlite/tables/SqliteTable.kt:
```diff
@@ -22,13 +22,6 @@ abstract class SqliteTable {
     @JvmField
     protected var columnIndex = 0
 
-    private var columnSet: SortedSet<Column>? = null
-    private var createTableSql: String? = null
-    private var preparedInsertSql: String? = null
-    private var preparedUpdateSql: String? = null
-    private var createIndexesSql: Collection<String>? = null
-    private var selectSql: String? = null
```
AFAIK these were never written to. Were we planning on using them as lazily computed values?

Lax125 commented on src/main/kotlin/com/zepben/evolve/database/sqlite/writers/MeasurementServiceWriter.kt:
```diff
@@ -7,6 +7,5 @@
-class MeasurementServiceWriter(hasCommon: (String) -> Boolean, addCommon: (String) -> Boolean) {
-
-}
+// TODO find out if this is still needed
+class MeasurementServiceWriter(hasCommon: (String) -> Boolean, addCommon: (String) -> Boolean)
```
The measurement service doesn't seem to have a corresponding consumer client or consumer service in ewb-network-routes. Do we still need it?

Lax125 replied:
I removed it. A proper implementation can be added once we support the measurement service in the database.

Lax125 commented on src/main/kotlin/com/zepben/evolve/services/diagram/DiagramServiceInstanceCache.kt:
```diff
@@ -10,7 +10,7 @@ package com.zepben.evolve.services.diagram
 import com.zepben.evolve.cim.iec61970.base.diagramlayout.DiagramObjectPoint
 
-@Suppress("UnstableApiUsage")
+// TODO find out if this is used
 class DiagramServiceInstanceCache {
```
There doesn't seem to be a usage for either `DiagramServiceInstanceCache` or `NetworkServiceInstanceCache` in Zepben's repos. Is this meant to be public API?

Lax125 replied:
Removed both classes.

Lax125 commented on src/main/kotlin/com/zepben/evolve/services/common/extensions/IdentifiedObjectExtensions.kt:
```diff
@@ -33,6 +29,7 @@ internal fun IdentifiedObject.validateReference(other: IdentifiedObject, getter:
 
 /**
  * Removes all names that belong to this object from their corresponding [NameType][com.zepben.evolve.cim.iec61970.base.core.NameType].
+ * TODO: investigate whether this is superseded by clearNames()
  */
 fun IdentifiedObject.removeNamesFromTypes() {
     names.forEach { it.type.removeName(it) }
```
What was the original purpose of this function that removes an io's names from their types but doesn't remove them from the io?

Lax125 replied:
It is used in the gis network extractor when the io is going to be unlinked and removed from the network anyway, so I suppose it's meant to be a faster(?) way to mark the io and its names as unreachable and garbage-collectable

Lax125 replied:
Removed due to limited usage.

Lax125 commented on src/test/kotlin/com/zepben/evolve/utils/ComparatorTestHelpers.kt:
```diff
@@ -16,8 +16,6 @@ import org.hamcrest.Matchers.*
 import kotlin.reflect.KMutableProperty1
 import kotlin.reflect.KProperty1
 
-data class ExpectedDifference(val name: String, val difference: Difference)
```
This was unused since its creation in all of our repos.

Lax125 commented:
There are several untested public methods and constructors (e.g. `Connect.connectWithIdentity`, `CustomerConsumerClient.constructor(ManagedChannel, CallCredentials?)`. I could try writing tests for all of them but it will take quite a while. Shall I mark them with FIXME's instead?

Lax125 commented:
What's the point of the `nameAndMRID()` extension functions for protobuf messages?

Lax125 commented:
`nameAndMRID()` removed for protobuf message types.

Lax125 commented on src/test/kotlin/com/zepben/evolve/services/network/testdata/FillFields.kt:
```diff
@@ -667,6 +667,9 @@ fun EquivalentEquipment.fillFields(service: NetworkService, includeRuntime: Bool
+
     return this
 }
 
```
Question: regarding the comment on `Switch.fillFIelds` (github won't let me comment on the exact lines sorry), do we plan on ever adding unganged switch state to the protobuf library?

charlta replied:
I would expect it to be added at some point, particularly if we get traction in the US where unganged is more common.

Lax125 commented:
There's also thousands of public/protected fields and methods without kdoc, but we'll be here all month if I go through them all

Lax125 commented on src/main/kotlin/com/zepben/evolve/database/sqlite/tables/TableMetadataDataSources.kt:
```diff
@@ -13,15 +13,15 @@ import com.zepben.evolve.database.sqlite.tables.Column.Nullable.NOT_NULL
-    val SOURCE = Column(++columnIndex, "source", "TEXT", NOT_NULL)
-    val VERSION = Column(++columnIndex, "version", "TEXT", NOT_NULL)
-    val TIMESTAMP = Column(++columnIndex, "timestamp", "TEXT", NOT_NULL)
+    val SOURCE: Column = Column(++columnIndex, "source", "TEXT", NOT_NULL)
```
Looks clunkier, but is recommended since the db code is public API.

Lax125 commented on src/main/kotlin/com/zepben/evolve/cim/iec61968/common/Location.kt:
```diff
@@ -24,18 +24,18 @@ class Location @JvmOverloads constructor(mRID: String = "") : IdentifiedObject(m
 
     /**
-     * Sequence of position points describing this location, expressed in coordinate system [Location.CoordinateSystem].
+     * Sequence of [PositionPoint]s describing this location.
```
I looked in the CIM docs and found `CoordinateSystem` to be [a real thing](https://zepben.github.io/evolve/docs/cim/cim100/TC57CIM/IEC61968/Common/CoordinateSystem/) that just isn't implemented in the SDK.

Lax125 commented on src/main/kotlin/com/zepben/evolve/services/common/extensions/IdentifiedObjectExtensions.kt:
```diff
@@ -14,10 +14,6 @@ import com.zepben.evolve.cim.iec61970.base.core.IdentifiedObject
 fun IdentifiedObject.nameAndMRID(): String = if (name.isBlank()) mRID else "'$name' [$mRID]"
 fun IdentifiedObject.typeNameAndMRID(): String = if (name.isBlank()) "${javaClass.simpleName} $mRID" else "${javaClass.simpleName} $name [$mRID]"
 
-internal fun <T : IdentifiedObject> Iterable<T>?.containsMRID(mRID: String): Boolean {
-    return getByMRID(mRID) != null
-}
```
Never used

Lax125 commented on src/main/kotlin/com/zepben/evolve/services/diagram/DiagramService.kt:
```diff
@@ -11,12 +11,6 @@ import com.zepben.evolve.cim.iec61970.base.core.IdentifiedObject
 import com.zepben.evolve.services.common.BaseService
-import kotlin.reflect.KClass
-
-private val supportedClasses: Set<KClass<out IdentifiedObject>> = setOf(
-    Diagram::class,
-    DiagramObject::class
-)
```
Never used

charlta replied:
I suspect this is left over from before `BaseService.supportedClasses` was added,

Lax125 commented on src/main/kotlin/com/zepben/evolve/services/network/NetworkServiceComparator.kt:
```diff
@@ -208,16 +211,6 @@ class NetworkServiceComparator @JvmOverloads constructor(var options: NetworkSer
 
-    /************ IEC61968 COMMON ************/
-
-    private fun compareLocation(source: Location, target: Location): ObjectDifference<Location> =
-        ObjectDifference(source, target).apply {
-            compareIdentifiedObject()
-
-            compareValues(Location::mainAddress)
-            compareIndexedValueCollections(Location::points)
-        }
```
Moved to `BaseServiceComparator`

Lax125 replied:
moved back, see comment below for why I moved it

Lax125 commented on src/main/kotlin/com/zepben/evolve/services/network/tracing/feeder/AssociatedTerminalTrace.kt:
```diff
@@ -22,13 +22,6 @@ import com.zepben.evolve.services.network.tracing.traversals.BasicTraversal
-    /**
-     * @return a traversal ignoring open points.
-     */
-    internal fun newTrace(): BasicTraversal<Terminal> {
-        return newTrace(OpenTest.IGNORE_OPEN)
-    }
```
Never used, since `AssignTo(Lv)Feeders` runs separate traces for normal and current network state. Can be added back in the future if required.

Lax125 commented on src/test/kotlin/com/zepben/evolve/utils/CimToProtoValidatorHelpers.kt:
```diff
@@ -7,27 +7,10 @@
     mrid?.let { assertThat(pb, equalTo(it)) } ?: assertThat(pb, emptyString())
 }
-
-internal fun validateMRID(pb: String, cim: IdentifiedObject?) {
-    validateMRID(pb, cim?.mRID)
-}
-
-internal fun validateMRIDList(pb: ProtocolStringList, cim: Collection<IdentifiedObject>) {
-    assertThat(pb.size, equalTo(cim.size))
-    if (cim.isNotEmpty())
-        assertThat(pb, containsInAnyOrder(*cim.stream().map { it.mRID }.toArray()))
-}
-
-internal fun validateMRIDList(pb: ProtocolStringList, cim: List<IdentifiedObject>) {
-    assertThat(pb.size, equalTo(cim.size))
-    if (cim.isNotEmpty())
-        assertThat(pb, contains(*cim.stream().map { it.mRID }.toArray()))
-}
```
All unused.

Lax125 commented on src/test/kotlin/com/zepben/evolve/streaming/RunStreaming.kt:
```diff
@@ -53,21 +51,6 @@ private fun runRetrieve(client: NetworkConsumerClient) {
     log("Num objects: ${client.service.num<IdentifiedObject>()}")
 }
 
-private fun runCustomersForFeeder(client: CustomerConsumerClient) {
-    val result = client.getCustomersForContainer(setOf(""))
-    result.throwOnError()
-    log("Num unresolved: ${client.service.numUnresolvedReferences()}")
-    log("Num objects: ${client.service.num<IdentifiedObject>()}")
-}
-
-private fun runDiagramObjects(client: DiagramConsumerClient) {
-    val result = client.getDiagramObjects(setOf(""))
-    result.throwOnError()
-    log("Num unresolved: ${client.service.numUnresolvedReferences()}")
-    log("Num objects: ${client.service.num<IdentifiedObject>()}")
-}
-
```
Unused

Lax125 commented on src/test/kotlin/com/zepben/evolve/services/network/NetworkServiceComparatorTest.kt:
```diff
@@ -210,25 +207,6 @@ internal class NetworkServiceComparatorTest : BaseServiceComparatorTest() {
-    /************ IEC61968 COMMON ************/
-
-    @Test
-    internal fun compareLocation() {
-        compareIdentifiedObject { Location(it) }
-
-        comparatorValidator.validateProperty(
-            Location::mainAddress,
-            { Location(it) },
-            { StreetAddress(townDetail = TownDetail("town", "state")) },
-            { StreetAddress(townDetail = TownDetail("other", "state")) })
-        comparatorValidator.validateIndexedCollection(
-            Location::points,
-            Location::addPoint,
-            { Location(it) },
-            { PositionPoint(1.0, 2.0) },
-            { PositionPoint(3.0, 4.0) })
-    }
```
Moved to `BaseServiceComparatorTest`

Lax125 replied:
Moved back

Lax125 commented on src/test/kotlin/com/zepben/evolve/services/common/UnknownPrimitivesTest.kt:
```diff
@@ -1,28 +0,0 @@
-
-    @Test
-    internal fun checkEqualsOperator() {
-        assertThat(UNKNOWN_DOUBLE == UNKNOWN_DOUBLE, equalTo(true))
-        assertThat(UNKNOWN_INT == UNKNOWN_INT, equalTo(true))
```
Equals operator does not need to be checked here: for any `val x: Double` or `val x: Int`, `x == x` is guaranteed.

charlta replied:
I suspect this was for coverage at some stage and agree it is no longer relevant

Lax125 commented:
What's the policy on what should be public in the test directory? Since tests aren't packaged into the library, I think anything that can be reusable (test networks, utils) in tests should remain public to encourage devs to write documentation for them, which will help future devs that use them.

charlta commented on src/main/kotlin/com/zepben/evolve/services/network/tracing/connectivity/ConnectivityResult.kt:
```diff
@@ -42,11 +42,7 @@ class ConnectivityResult private constructor(
-        if (nominalPhasePaths != other.nominalPhasePaths) return false
-
-        return true
+        return fromTerminal == other.fromTerminal && toTerminal == other.toTerminal && nominalPhasePaths == other.nominalPhasePaths
```
nit: Prefer clarifying parentheses

Lax125 replied:
Added clarifying parentheses.

charlta commented on src/main/kotlin/com/zepben/evolve/testing/TestNetworkBuilder.kt:
```diff
@@ -21,7 +21,7 @@ import kotlin.reflect.full.primaryConstructor
  * A class for building simple test networks, often used for unit testing.
  */
-open class TestNetworkBuilder {
+internal open class TestNetworkBuilder {
```
urgent: This must remain a public class as it is very widely used by migrators. It was deliberately made `open` to allow people that want to add CIM extension classes to extend this builder to help them build their own object, but that may have been alleviated by the improvements to `fromOther` etc, so happy for that to be removed if you insist.

Lax125 replied:
Removed `internal`.

charlta commented on src/test/kotlin/com/zepben/evolve/services/common/ServiceDifferencesTest.kt:
```diff
@@ -16,14 +16,14 @@ import org.hamcrest.Matchers.*
     @JvmField
     @RegisterExtension
-    var systemOutRule: SystemLogExtension = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess()
+    var systemOut: SystemLogExtension = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess()
```
todo: Should have been renamed to `systemErr`

Lax125 replied:
Renamed as such.

Lax125 replied:
Renamed as such.

charlta commented on src/test/kotlin/com/zepben/evolve/services/network/tracing/phases/PhaseStepTest.kt:
```diff
@@ -17,14 +17,14 @@ import org.junit.jupiter.api.extension.RegisterExtension
     @JvmField
     @RegisterExtension
-    var systemOutRule: SystemLogExtension = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess()
+    var systemOut: SystemLogExtension = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess()
```
todo: should have been renamed `systemErr`

Lax125 replied:
Renamed as such.

charlta commented on src/test/kotlin/com/zepben/evolve/services/network/tracing/FindWithUsagePointsTest.kt:
```diff
@@ -26,11 +26,11 @@ import org.junit.jupiter.api.extension.RegisterExtension
     @JvmField
     @RegisterExtension
-    var systemOutRule: SystemLogExtension = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess()
+    var systemOut: SystemLogExtension = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess()
```
todo: wont comment on any more, but all `SystemLogExtension.SYSTEM_ERR` should become `systemErr` not `systemOut`

Lax125 replied:
Renamed as such. This should be the last one.

Lax125 commented on src/test/kotlin/com/zepben/evolve/streaming/mutations/SwitchStateClientTest.kt:
```diff
@@ -9,57 +9,120 @@
+        val switchToUpdate = SwitchStateUpdate("id", true)
+        val result = SwitchStateClient(grpcChannel).setCurrentSwitchState(switchToUpdate)
+
+        assertThat(result.wasSuccessful, equalTo(true))
```
Fixed for this test, will fix for others before re-review. I'll make an exception for boolean fields for CIM class tests and their corresponding columns in the changelog tests, since we are checking the data held in those fields/columns rather than the success/failure of a process. This also makes it consistent with other field/column checks.

Lax125 replied:
Wrote assertion messages for the other boolean assertions.

Lax125 commented on src/test/kotlin/com/zepben/evolve/streaming/mutations/SwitchStateClientTest.kt:
```diff
@@ -9,57 +9,120 @@
     var systemOut: SystemLogExtension = SystemLogExtension.SYSTEM_OUT.captureLog().muteOnSuccess()
 
-    private val stub = mock(SwitchStateServiceGrpc.SwitchStateServiceBlockingStub::class.java)
+    private val stub: SwitchStateServiceGrpc.SwitchStateServiceBlockingStub = mockk(relaxed = true)
```
Changed as such, and also removed the type annotation for `client` since it's also private and it's pretty obvious that it's a `SwitchStateClient` from its definition.

Lax125 commented:
Regarding why I moved `compareLocation`, I was in the middle of working out why `compareOrganisation` was marked as unused and my brain thought that location should also belong in the `BaseServiceComparator` becuase it's in `com.zepben.evolve.cim.iec61968.common`. I've moved both the comparison function and its test back to their original places.

Lax125 commented on src/main/kotlin/com/zepben/evolve/services/common/BaseServiceComparator.kt:
```diff
@@ -116,7 +116,7 @@ abstract class BaseServiceComparator {
     }
 
-    protected fun compareNameType(source: NameType, target: NameType): ObjectDifference<NameType> =
+    private fun compareNameType(source: NameType, target: NameType): ObjectDifference<NameType> =
```
Used in `compare(BaseService, BaseService)`, which is inherited

Lax125 commented on src/main/kotlin/com/zepben/evolve/services/common/BaseServiceComparator.kt:
```diff
@@ -183,7 +185,7 @@ abstract class BaseServiceComparator {
     }
 
-    fun <T : IdentifiedObject> ObjectDifference<T>.compareNames(
+    private fun <T : IdentifiedObject> ObjectDifference<T>.compareNames(
```
Used in `compareIdentifiedObject()`, which is inherited

Lax125 commented on src/main/kotlin/com/zepben/evolve/services/common/BaseServiceComparator.kt:
```diff
@@ -28,7 +28,7 @@ import kotlin.reflect.jvm.isAccessible
 
     @Suppress("UNCHECKED_CAST")
-    val compareByType: Map<KType, KFunction<ObjectDifference<*>>> = this::class.memberFunctions
+    private val compareByType: Map<KType, KFunction<ObjectDifference<*>>> = this::class.memberFunctions
```
Used in both `compare` methods, which are inherited

Lax125 commented on src/test/kotlin/com/zepben/evolve/cim/iec61968/customers/CustomerTest.kt:
```diff
@@ -23,13 +23,13 @@ internal class CustomerTest {
     @Test
     internal fun hasEndDevicesReflectsNumEndDevices() {
         val customer = Customer()
-        assertThat("Customer has no end devices by default", not(customer.hasEndDevices()))
```
This was using the `assertThat(actual: T, matcher: Matcher<T>)` method, which would pass even if `customer.hasEndDevices()` returned true. Also fixed other cases of this throughout the repo.

Lax125 commented on src/test/kotlin/com/zepben/evolve/services/network/NetworkServiceTest.kt:
```diff
@@ -100,108 +100,129 @@ internal class NetworkServiceTest {
-        val node3 = createNodeForConnecting(network, "node3", 2)
-        val node4 = createNodeForConnecting(network, "node4", 2)
-        val node5 = createNodeForConnecting(network, "node5", 2)
+        val j0 = createJunctionForConnecting(network, "j0", 2)
```
I found it a bit confusing to name junctions node0/node1/etc since we also have connectivity nodes in a node breaker model, so I renamed all of them to j0/j1/etc. I left `StupidlyLargeNetwork` untouched as per the contribution doc.

Lax125 commented:
Will un-draft after implementing `t1`, `t2`, `t3` helpers to `ConducitngEquipment`, tests for said helpers, and using them where appropriate

kgreav commented:
awesome work @Lax125, didn't realise there was that much mess in the SDK still :grimacing: significant improvement.

can someone tell me though - was there anything in here that needs to be reflected in the Python SDK?

Lax125 commented:
@kgreav the `t1`, `t2`, `t3` helpers, the removed functions if they exist in the Python SDK, and possibly some typos for copied-over documentation. Other cleanup stuff was specific to problems found in the kotlin codebase. Pylint should help if we go for a cleanup of that Python SDK as well.
